### DISCUSSION TITLE
How to access reader endpoint correctly?

### DISCUSSION THREAD

**Discussion by jiyanhbo** [aws/aws-advanced-jdbc-wrapper#474] [Q&A]
I followed the code sampler to execute read & write sql statement by calling reader & write endpoints. However, the read requests are not split evenly at replica instances.  The read endpoint has 3 replica instances and most of db connection for reading operations come from a single instance. 

The metric for reader db connection can be found here (instance _1 is primary instance and others are replica ones)

<img width="428" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/3e069b9b-43db-403d-91d6-63a2d736aff3">

Do I set up anything wrong?  My understanding is that the wrapper will 
1) parse read endpoint to figure out all replica hosts (here is 3 hosts)
2) create db connection pools for each replica instance
3) evenly split reading operations to each replica's connection pool

Is that correct?  if yes, Should I add connection pool initiation code explicitly to fix the issue? if not, how to split the read traffic evenly among replica instances?

**Comment by jiyanhbo** [comment]
To be more specific, I followed the code https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/main/examples/AWSDriverExample/src/main/java/software/amazon/ReadWriteSplittingPostgresExample.java where internal connection pool is commented.  

I will  comment out the section to use internal connection pool and check if the above issue is addressed or not

**Reply by jiyanhbo** [reply]
I added back the connection pool but still see similar kind of db connection issue. Most of connection are from one postgres instance.  Any clue? 

<img width="558" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/d3942a1f-6b16-411d-95a2-ba1bffa8ed8f">

**Comment by congoamz** [comment]
Hi @jiyanhbo, thank you for reaching out with this problem. I had a few questions that will help clarify the problem for me.
1. Are you connecting directly to a reader or are you connecting to the writer and then calling setReadOnly(true)?
2. What does your URL look like when you are establishing the reader connections? Is it the standard aurora writer or reader cluster or something else?

**Reply by jiyanhbo** [reply]
I connect with reader directly.  

The url is "jdbc:postgresql:aws://foo-beam-amer-aurpg-virginia-dev-v1.cluster-ro-cnssohswqu7e.us-east-1.rds.amazonaws.com:5432/foo-beam-amer", It is standard aurora cluster's reader endpoint.

**Comment by jiyanhbo** [comment]
The cluster topology can be found 
<img width="797" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/8b5fe618-c820-4778-b6a4-190792df4f54">

The metrics for db connections can be found here
<img width="702" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/ae9096f6-54b2-49fc-b9a1-f734d7d0429f">

You can see that most of db connections are from one postgres instance.

**Reply by jiyanhbo** [reply]
I increased the tops by 5X and expect the db connections from two idle instances 1 & 3 should bump up. However, they are still super small and instance 2's connection jumped a lot.  

<img width="614" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/9d81ef3b-0b01-4ad2-ad61-95082e8c6167">

My requirement is very simple, the db connection or read traffic should be evenly split among all reader instances.

**Reply by jiyanhbo** [reply]
I stopped the reader traffic and restarted later, this time most of db connections are from instance 1. In the previous test, it is from instance 2
<img width="496" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/3f05e3ac-15d0-4cc7-880c-28fb183b1ceb">

Also I set the max connection pool size is 10 at HikariConfig. There are 3 service nodes, so Max db connections to a postgres replica instance should be 30.  However, we can see much more connections than the limit in AWS metrics. Looks like the thread pool max limit is not working as expected.

**Comment by congoamz** [comment]
Hi @jiyanhbo, I believe I understand what is happening here. There are two ways to get reader load balancing:
1. When calling setReadOnly(true) on a writer connection, our driver will connect to a reader if the read-write plugin is loaded. In this scenario, new reader connections are load-balanced. Note that the load balancing strategy is random reader selection, so load-balancing will not be perfectly even but should be mostly even
2. When connecting to the reader cluster endpoint, Aurora will load balance which reader you connect to

In your situation, you are load-balancing readers using the 2nd strategy (connecting to the reader cluster endpoint). Note that this method of load-balancing is completely controlled by Aurora, not our driver. I found [this Aurora documentation](https://repost.aws/knowledge-center/aurora-mysql-postgresql-reader-nodes) that mentions some reasons why you might not be seeing load-balancing when connecting to the reader cluster. From my own testing, I realized that Aurora will connect you to the same reader if you are establishing many connections within a very short time frame. If you wait 5 seconds between each connection (as shown in that Aurora doc page I linked), you should see more load-balancing happening. If not, you may want to review the other reasons on that page on why you might not be seeing load-balancing.

Regarding the connection pool size not being obeyed, please note that initial connections to a cluster URL will not be pooled. This is intentional as pooling cluster URLs may be problematic because they resolve to different instances over time. The instances are not capped at 30 in your example because the connections are using the cluster URL and thus are not being pooled. Currently, the main benefit of internal pools is when setReadOnly is called. When setReadOnly is called (regardless of the initial connection URL), an internal pool will be created for the writer/reader that the plugin switches to and connections for that instance can be reused in the future.

**Reply by renatomariscal** [reply]
Hi @congoamz, does `aws-advanced-jdbc-wrapper` takes care of proper DNS caching (within its power, not considering possible OS and like caching), such as managing/overriding Java's DNS cache TTL? Since aurora load balancing is based on DNS, that could be a factor.

**Reply by jiyanhbo** [reply]
Hi @congoamz, I followed the first way to get reader load balancing.  The load balancing has some improvement but still there is some weird behavior in the db connections metric after the change.  One replica instance does not have any db connections and other two have db connections which is close to be evenly distributed. Any clue about why one instance has no connection at all? 

<img width="416" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/6f29ce14-7c15-4ab1-8e36-77f676733d96">


Also can you confirm if we really need to call setReadonly(true) on a writer connection? I think call setReadonly(true) on a reader connection should also work.  [read-write plugin](https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/main/docs/using-the-jdbc-driver/using-plugins/UsingTheReadWriteSplittingPlugin.md#supplying-the-connection-string) says any url that can connect to the db should work and setReadonly(true) should create a physical connection and pool it.

**Reply by congoamz** [reply]
Hi @jiyanhbo, that is strange that the one instance has no connections. I tried to reproduce the issue on my end and was unable to do so. Can you try running your code again, and if you are still running into the problem, are you able to send me sample code that is triggering the issue on your end?

You do not need to call setReadOnly(true) on a writer connection unless you want the reader connection to be internally pooled. With the current implementation we do not pool initial connections to a cluster URL because they can resolve to different instances over time. Internal connection pooling is specifically intended to improve setReadOnly performance. You can use the reader cluster endpoint for the initial connection (will be load balanced by Aurora but not internally pooled) or connect to the writer and then call setReadOnly(true) (will be internally pooled and load-balanced but requires the extra connection to the writer first). To clarify the documentation you are referring to, you can use any URL to connect and read-write split. In the case of cluster URLs the initial connection will not be pooled, but the connection is still usable and you can still read-write split when you switch between the writer/reader via calls to `setReadOnly`.

**Reply by congoamz** [reply]
@renatomariscal the driver does not manage/override the DNS cache TTL

**Reply by jiyanhbo** [reply]
@congoamz, I found one issue in my code. The original connection string miss the prefix `aws-wrapper`. I added it back and the current connection string is:  "jdbc:aws-wrapper:postgresql://foo-beam-amer-aurpg-virginia-dev-v1.cluster-ro-cnssohswqu7e.us-east-1.rds.amazonaws.com:5432/foo_beam_amer" which is the endpoint url of the reader cluster

My code logic strictly follows the [example for read-write split](https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/247c28e990478390e32f2d55d0d9ad6d8cdfe80b/examples/AWSDriverExample/src/main/java/software/amazon/ReadWriteSplittingPostgresExample.java) except that connection pool is enabled in my code.

For connection properties, I set as below
    // Enable readWriteSplitting, failover, and efm plugins and set properties
    props.setProperty(PropertyDefinition.PLUGINS.name, "readWriteSplitting,failover,efm");
    props.setProperty(PropertyDefinition.USER.name, USERNAME);
    props.setProperty(PropertyDefinition.PASSWORD.name, PASSWORD);

but now I hit an error `An error occurred while obtaining the connections host ID.`. I am still investigating what happened. Do you have any clue?

Here is the error stack
`
software.amazon.jdbc.hostlistprovider.AuroraHostListProvider.identifyConnection(AuroraHostListProvider.java:671), software.amazon.jdbc.PluginServiceImpl.identifyConnection(PluginServiceImpl.java:502), software.amazon.jdbc.PluginServiceImpl.fillAliases(PluginServiceImpl.java:531), software.amazon.jdbc.plugin.efm.HostMonitoringConnectionPlugin.connectInternal(HostMonitoringConnectionPlugin.java:282), software.amazon.jdbc.plugin.efm.HostMonitoringConnectionPlugin.connect(HostMonitoringConnectionPlugin.java:271), software.amazon.jdbc.ConnectionPluginManager.lambda$connect$4(ConnectionPluginManager.java:375), software.amazon.jdbc.ConnectionPluginManager.lambda$makePluginChainFunc$2(ConnectionPluginManager.java:280), software.amazon.jdbc.ConnectionPluginManager.lambda$null$1(ConnectionPluginManager.java:280), software.amazon.jdbc.plugin.staledns.AuroraStaleDnsHelper.getVerifiedConnection(AuroraStaleDnsHelper.java:66), software.amazon.jdbc.plugin.failover.FailoverConnectionPlugin.connectInternal(FailoverConnectionPlugin.java:730), software.amazon.jdbc.plugin.failover.FailoverConnectionPlugin.connect(FailoverConnectionPlugin.java:723), software.amazon.jdbc.ConnectionPluginManager.lambda$connect$4(ConnectionPluginManager.java:375), software.amazon.jdbc.ConnectionPluginManager.lambda$makePluginChainFunc$2(ConnectionPluginManager.java:280), software.amazon.jdbc.ConnectionPluginManager.lambda$null$1(ConnectionPluginManager.java:280), software.amazon.jdbc.plugin.readwritesplitting.ReadWriteSplittingPlugin.connectInternal(ReadWriteSplittingPlugin.java:132), software.amazon.jdbc.plugin.readwritesplitting.ReadWriteSplittingPlugin.connect(ReadWriteSplittingPlugin.java:127), software.amazon.jdbc.ConnectionPluginManager.lambda$connect$4(ConnectionPluginManager.java:375), software.amazon.jdbc.ConnectionPluginManager.lambda$makePluginChainFunc$2(ConnectionPluginManager.java:280), software.amazon.jdbc.ConnectionPluginManager.executeWithSubscribedPlugins(ConnectionPluginManager.java:258), software.amazon.jdbc.ConnectionPluginManager.connect(ConnectionPluginManager.java:372), software.amazon.jdbc.wrapper.ConnectionWrapper.init(ConnectionWrapper.java:135), software.amazon.jdbc.wrapper.ConnectionWrapper.<init>(ConnectionWrapper.java:88), software.amazon.jdbc.Driver.connect(Driver.java:126), java.sql/java.sql.DriverManager.getConnection(DriverManager.java:681), java.sql/java.sql.DriverManager.getConnection(DriverManager.java:190), com.wbdstreaming.foo.FooController.readPostgres(FooController.java:62),
`

**Reply by jiyanhbo** [reply]
The error is from plugin efm. I removed it and now all query work. So maybe there is a bug from efm.

I still see the db connections not evenly distributed. Below is the core logic of the controller class `FooController` of my app that access postgres db. 

The connection pool init method is copied from the existing sample code. I updated the pool key as a combination of host + username which should be fine

The way to access postgres readers is a combination of two options you suggested
1) call reader endpoint of cluster
2) call setReadOnly(true) on connection.

Do you think the connection pool should work?

```
    private static HikariConfig getHikariConfig(HostSpec hostSpec, Properties props) {
        final HikariConfig config = new HikariConfig();
        config.setMaximumPoolSize(10);
        config.setInitializationFailTimeout(75000);
        config.setConnectionTimeout(10000);
        return config;
    }

    private static String getPoolKey(HostSpec hostSpec, Properties props) {
        // Include the URL and user in the connection pool key so that a new
        // connection pool will be opened for each different instance-user
        // combination.
        final String user = props.getProperty(PropertyDefinition.USER.name);
        return hostSpec.getUrl() + user;
    }

  /// **this method is called at the constructor of the controller class** 
   public  static void initConnectionPool(){
        final HikariPooledConnectionProvider connProvider =
                new HikariPooledConnectionProvider(
                        FooController::getHikariConfig,
                        FooController::getPoolKey
                );
        ConnectionProviderManager.setConnectionProvider(connProvider);
    }

    public ResponseEntity<String> readPostgres(@RequestBody String query)  {
        Properties roProperties = new Properties();
        roProperties.setProperty(PropertyDefinition.PLUGINS.name, "readWriteSplitting");
        roProperties.setProperty(PropertyDefinition.USER.name, postgresConfig.getROUsername());
        roProperties.setProperty(PropertyDefinition.PASSWORD.name, postgresConfig.getROPassword());

        String result;
        try (final Connection conn = DriverManager.getConnection(
                "jdbc:aws-wrapper:postgresql://foo-beam-amer-aurpg-virginia-dev-v1.cluster-ro-cnssohswqu7e.us-east-1.rds.amazonaws.com:5432/foo_beam_amer",
                roProperties)) {
            PostgreUtil.setInitialSessionSettings(conn);

            // Begin business transaction
            conn.setAutoCommit(false);
            result = PostgreUtil.executeWithFailoverHandling(conn, query, true);

            // Commit business transaction
            conn.commit();
            // Internally switch to a reader connection
            conn.setReadOnly(true);
        }
       .... 
}
```

**Reply by jiyanhbo** [reply]
Here is the db connections metrics

<img width="476" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/ef34e529-5dce-443d-b0a6-594e0c8cfa13">

You can see that db connections fluctuates at every reader instance, which cause the cpu utilization fluctuate a lot also. That looks weird. @congoamz ,did you see the similar kind of fluctuate pattern in your test or can you post a metrics for db connections from your test? Is there a way to confirm if the wrapper's connection pool is really working in my code?

**Reply by jiyanhbo** [reply]
Look at the cpu utilization metrics from one reader instance. You can see that it fluctuates
a lot which is the major concern from ours. We expect it is very flat and stable. 

<img width="370" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/32f52f17-2a3e-43df-9102-81e766c74575">

**Reply by congoamz** [reply]
Hi @jiyanhbo, I have a couple recommendations/clarifications:
- your setting for the plugin property should be either `auroraHostList,readWriteSplitting` or `readWriteSplitting,failover`. The auroraHostList and failover plugins inform the driver that an Aurora cluster is being used and that it should query for Aurora's topology. More information can be found [here](https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/main/docs/using-the-jdbc-driver/using-plugins/UsingTheReadWriteSplittingPlugin.md#loading-the-read-write-splitting-plugin)
- with your current code you are connecting to a reader via the reader cluster endpoint (which will not be pooled) and then calling setReadOnly(true). Since the r/w plugin knows you are already connected to a reader, it does not connect you to a new instance when setReadOnly(true) is called. Consequently, no internal pools are opened and the specific instances that you are connecting to are determined by the reader cluster endpoint and not the driver. If instead you connected to the writer cluster and then called setReadOnly(true), the r/w plugin would open a pooled connection to a reader, and this reader selection would be load-balanced by the driver. Note that the writer connection is cached and kept open by the r/w plugin in this scenario so that it can quickly switch back to the it if setReadOnly(false) is called
- as mentioned in the point above, the driver is not opening internal pools with your current code. I could be wrong but I believe the graph you are seeing is fluctuating because connections are regularly being opened and then quickly closed and there is no pooling

**Reply by jiyanhbo** [reply]
Thanks! Before I set the plugins as "readWriteSplitting,failover,efm" but it throw the error that I posted above. Let me try "readWriteSplitting,failover".

**Reply by jiyanhbo** [reply]
Just confirmed that connecting to writer cluster and call setReadOnly(true) will trigger connection pool for reader instances. But the connection pool looks like not reused. From metrics, I can see that the connection pool to reader cluster are generated per configuration. But all reader instances have very small tps (around 2) while writer instance have super high tps (60) and cpu utilization. The connection pool for reader instances should not be used. 

@congoamz , do you have any clue why the connection pool is created but not reused?

**Reply by jiyanhbo** [reply]
I go through AWS wrapper code and realized  why the connection pool is created but not used. 

`
try (final Connection conn = DriverManager.getConnection(
                "jdbc:aws-wrapper:postgresql://foo-beam-amer-aurpg-virginia-dev-v1.cluster-ro-cnssohswqu7e.us-east-1.rds.amazonaws.com:5432/foo_beam_amer",
                roProperties)) {
`

In the above code, a writer connection object is created firstly for every incoming request. The connection object is switched to a reader connection object later when setReadonly(true) is called. However, the connection object is never used later after the try catch statement.  That is why connection pool is created but not used and all reading traffic go to writer instance. 


So I tried to initiate  a reader connection object when the service starts and reuse it for all requests. However, the reader object always connects to  a specific reader instance and reading traffic is not load balanced as expected.

Here is db connection & tps for reader foo-beam-amer-aurpg-virginia-dev-v1-2
<img width="350" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/d8fbe837-e980-4103-a0ae-c50036612bd0">

<img width="350" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/9d730695-d283-415c-8ae4-f7426733a0f8">

Here is db connection & tps for reader instance foo-beam-amer-aurpg-virginia-dev-v1-3
<img width="356" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/fbd991c3-5b73-427b-9b38-158ee446d5c9">

<img width="347" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/59489b34-965a-45e9-9acc-a6842f4990a2">

Here is db connection & tps for reader instance foo-beam-amer-aurpg-virginia-dev-v1-4
<img width="360" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/f7c90356-9e14-4ce4-bea1-6532f8df9967">

<img width="350" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/a2266be1-6ee1-48e7-ab91-c3418e31333c">

So there are few problems here
1.  one reader instance has very few tps. It looks like connection pool is not created for it
2.  Other two reader instances have stable db connections no matter how much the tps is. The connection pool should be created for them. But why the connections are 15 & 25 separately? Based on my configuration, both should be around 30.  

From metrics, we can see connection pool is working but load balance does not. Here is my understanding why load balance does not work. When the read-writer split plugin call setReadonly(true) and current connection is writer, it will
1. choose a random avaialbe reader instance from topology and create a new reader connection
2. switch the current writer connection with the reader connection

when we reuse the connection wrapper, it will connect to the same reader instance always. So  the current solution should not load balance reading traffic well.   

@congoamz, please help confirm if my understanding is correct or not.

**Reply by congoamz** [reply]
Hi @jiyanhbo, thanks for the feedback. Had a few comments/questions:
- yes, the first time that setReadOnly(true) is called on a writer connection the driver will connect to a random reader instance. So reader connections established by setReadOnly specifically should be roughly load-balanced. It will not be perfectly load balanced because the selection strategy is random. A least-connections selection strategy for internal pools was added to the driver code recently, which should result in more even load-balancing across readers, however this is not in an official release yet. Couple of extra things to note here:
  - as you have observed, after setReadOnly(true) has been called for the first time on a writer connection, the driver will internally cache the new reader connection and the original writer connection. Calling setReadOnly(false), then setReadOnly(true) again on the same ConnectionWrapper object will not result in any new connections being opened. This was done to improve performance, since caching the connections saves us from having to open new connections every time setReadOnly is called on that ConnectionWrapper instance.
  - the r/w plugin load-balances connections created from calling setReadOnly but is not responsible for load-balancing otherwise. For example, if you connect directly to the Aurora reader cluster, load-balancing will be handled by Aurora and not by our driver, and if you connect directly to a reader instance URL, our driver will obey that request and connect you to that instance specifically
- I'm not sure if the reader with low TPS indicates that no connection pool was created for it. It may be that a connection pool was created for it but that specific reader was not selected often by the random selection strategy

Some questions:
- can you help me understand what you are trying to achieve by initiating a reader connection object when the service starts and reusing it for all requests? Are you doing this to try and avoid having to open a writer connection before triggering the internal pool with setReadOnly(true)? Is your goal to pool all connections (eg direct connections to the reader cluster in addition to reader connections opened by the r/w plugin on setReadOnly calls)?
- so far, it sounds like you have been executing test code to see how the driver behaves. In your actual application, what requirements, expectations, and goals do you have with the driver? If we have more info we may be able to offer suggestions for alternative configurations/settings that fit your use case better

**Reply by jiyanhbo** [reply]
I created the connection wrapper always by connecting with writer cluster firstly and then call setReadonly(true). So this connection wrapper will work as a reader connection. 

I tried two ways to use the connection wrapper. Here is the summary
1.  Create the connection wrapper when service starts and reuse it for all reading query. 
pros: connection pool is created and works at most of reader instances if they are connected by any service pod. Metrics shows the connection number & cpu utilization of every reader instance are stable. 
cons: reading traffic is not evenly distributed among reader instances. The reading traffic of a service pod connects to same reader instance. 

2. Create a connection wrapper for each incoming request. 
pros: reading traffics is evenly distributed among reader instances. 
cons: connection pool is not working as expected (I believe that the connection wrapper is not reused which cause the connection pool created for each connection wrapper is not used at all). Metrics shows that db connection number & cpu utilization fluctuates a lot at each reader instance

My expectation for the postgres lib is:
1. can distributes traffics of each service pod evenly among reader instance
2. each db instance should have very stable connection number and cpu utilization
3. can handle failover

From my testing, 1 & 2 do not work at the same time. 

To answer your question why I try to reuse connection wrapper, it is just because other ways to use connection wrapper do not generate expected result. I prefer to create connection wrapper for each request as all sample projects do. In this way, traffic is distributed evenly among db instances.  But the connection pool does not work as expected. The reason could be that connection wrapper objects created for different requests do not share a connection pool. 

 The issue may be addressed by sharing connection pool between connection wrapper.   @congoamz , let me know if my understanding is correct. If you think there is a way to address the issue without any code change, please help share the pseudo code so that I can test it.  Thanks for your help!

**Reply by congoamz** [reply]
Hi @jiyanhbo thanks for that info. You should not need to reuse the connection wrapper. The code already shares the internal connection pools between different connection wrapper objects. So even if you close a ConnectionWrapper, a future ConnectionWrapper will be able to use the pool established by the first ConnectionWrapper. I ran a test on my side and the TPS, connections, and CPU utilization all look stable and reasonable, so we will have to figure out why you are getting different results. I posted my test code and results below. Each reader instance has 10 connections (which is the default maximumPoolSize setting in hikari), CPU is stable around ~2%, and TPS is stable ~3.4 queries per second. Note that DatabaseConnections (in CloudWatch) seems to be a more reliable metric than Connection utilization (in performance insights). Here are some questions that may help us figure out why you are getting different results:
- do you notice any obvious differences between my test code and yours that may cause the different results?
- is the code and SQL you are executing just simple test code or is it real-life code that your application would run in a real-life scenario?
- what version of the driver are you using?
- in your original code snippet you were not executing any SQL after switching to the reader. Is that still the case? Are the SQL queries you are executing simple/fast or complex/slow?
- one thing I cannot tell from your original snippet is how often you are creating connections. Are you opening multiple connections rapidly in a loop? Do you have multiple connections open at the same time? Are you opening any connections in your code other than what is shown in your original snippet?

Test code:
```java
public static void main(String[] args) throws SQLException {
    // Connect to the writer cluster URL initially
    String url = "jdbc:aws-wrapper:postgresql://my-postgres-database.cluster-XYZ.us-east-2.rds.amazonaws.com/mydb";
    String user = "username";
    String password = "password";

    final Properties props = new Properties();
    props.setProperty(PropertyDefinition.USER.name, user);
    props.setProperty(PropertyDefinition.PASSWORD.name, password);
    props.setProperty(PropertyDefinition.PLUGINS.name, "auroraHostList,readWriteSplitting");

    ConnectionProviderManager.setConnectionProvider(
        new HikariPooledConnectionProvider((hostSpec, connProps) -> new HikariConfig()));

    try {
      while (true) {
        // Connect to writer
        try(Connection conn = DriverManager.getConnection(url, props)) {
          // Execute some arbitrary SQL
          final Statement writerStmt = conn.createStatement();
          writerStmt.execute("SELECT aurora_db_instance_identifier()");

          // Internally switch to reader
          conn.setReadOnly(true);

          // Execute some arbitrary SQL
          final Statement readerStmt = conn.createStatement();
          readerStmt.execute("SELECT aurora_db_instance_identifier()");
        }
      }
    } catch (SQLException e)  {
      ConnectionProviderManager.releaseResources();
      ConnectionProviderManager.resetProvider();
    }
  }
```
Database connections (orange line is writer, blue line is average connections across the cluster, brown line is a reader and overlaps all other reader lines):
![database-connections](https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/75754763/b395ee8c-c2ab-4cff-adf9-f8879ea4725d)

CPU utilization:
![cpu-utilization](https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/75754763/289b1b01-5bfe-4571-9303-53c13ef9a8f6)

TPS:
![queries-per-second](https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/75754763/151f6b4d-5a28-4d85-b880-ae36eba16166)

**Reply by jiyanhbo** [reply]
My code is almost same as your code.  I create a connection to writer instance and call setReadonly(true) immediately without running a random sql as what your code did. 

```
        // Connect to writer
        try(Connection conn = DriverManager.getConnection(url, props)) {
          // Internally switch to reader
          conn.setReadOnly(true);

          // Execute some arbitrary SQL
          final Statement readerStmt = conn.createStatement();
          readerStmt.execute("SELECT aurora_db_instance_identifier()");
        }
```

You can reproduce my issue when testing under high throughput. Under low throughput(<5), my test result is good also. But under high throughput(>50), the result is not as expected. 

My test is done through Jmeter with 100-200 users calling my service api concurrently. It cause tps around 100 normally. 

My code is from a read world app, not for test purpose only. The driver version is latest. 

Also I write a simple library to load balance the traffic by creating a map of reader instance and connection pool. For each request, I randomly select a reader instance & get a connection from its connection pool. The result looks very good and exactly what I expected.  I still think there is something wrong from the connection pool sharing. Can you point me where is the connection pool sharing logic from?

**Reply by congoamz** [reply]
Thanks for the info. How are you generating the higher throughput? I'd like to reproduce this on my end, so if you can provide full sample code to reproduce or if you can tell me how I should tweak my example to get the higher throughput that would be great. 

The code for the connection pools is [here](https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/61a8f2c365ab626ff62c5f39f4ef7be74d875bbf/wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java)

Given your sample code and the fact that we do not internally pool cluster URLs, I'm wondering if there's a better driver configuration for your workflow. The internal pools are meant to improve performance when switching often between the writer/reader in multiple different connections but it looks like you only need a reader in this case. Would it work for you to keep internal connection pools disabled and instead maintain your own external pool to the reader cluster? It looks like you are connecting to the writer before the reader specifically to trigger the internal connection pooling and using your own external pool would allow you to skip this step. You could also maintain a second external pool for the writer cluster if you'd like. Let me know if you think that configuration would work for you or not.

**Reply by jiyanhbo** [reply]
I used Jmeter for testing.  Here is how I configure it
<img width="573" alt="image" src="https://github.com/awslabs/aws-advanced-jdbc-wrapper/assets/64919660/dee737a7-253b-4d4d-970b-afe2200b029b">

You can use JMeter to do the same testing.

I will use connection pool for both reader & writer to make sure the postgres db instances in stable status. Let me know if you saw similar metrics pattern as what I saw.

**Reply by renatomariscal** [reply]
@jiyanhbo Are you using `AuroraHostListProvider` plugin?

**Reply by congoamz** [reply]
@jiyanhbo couple questions in addition to renatomariscal's above:
- I personally haven't used JMeter before, will I get the same behavior if I spin up 500 threads and have each thread execute my sample code in an infinite loop?
- in your code that you are running with Jmeter, are all 100-200 users executing the same SQL query?
